src/wasm/wasm-s-parser.cpp

```diff
@@ -2784,36 +2758,89 @@ void SExpressionWasmBuilder::parseInnerElem(Element& s,
-    throw ParseException("invalid inline function signature", s.line, s.col);
-  }
-  std::vector<Type> params;
-  std::vector<Type> results;
-  for (size_t k = 1; k < s.size(); k++) {
-    Element& curr = *s[k];
-    if (elementStartsWith(curr, PARAM)) {
-      auto newParams = parseParamOrLocal(curr);
-      params.insert(params.end(), newParams.begin(), newParams.end());
-    } else if (elementStartsWith(curr, RESULT)) {
-      auto newResults = parseResults(curr);
-      results.insert(results.end(), newResults.begin(), newResults.end());
+HeapType SExpressionWasmBuilder::parseHeapType(Element& s) {
+  if (s.isStr()) {
+    // It's a string.
+    if (s.dollared()) {
+      auto it = typeIndices.find(s.str().str);
+      if (it == typeIndices.end()) {
+        throw ParseException("unknown dollared function type", s.line, s.col);
+      }
+      return types[it->second];
+    } else {
+      // index
+      size_t offset = atoi(s.str().c_str());
+      if (offset >= types.size()) {
+        throw ParseException("unknown indexed function type", s.line, s.col);
+      }
+      return types[offset];
+    }
+  }
+  // It's a list.
+  if (*s[0] == FUNC) {
+    std::vector<Type> params;
+    std::vector<Type> results;
+    for (size_t k = 1; k < s.size(); k++) {
+      Element& curr = *s[k];
+      if (elementStartsWith(curr, PARAM)) {
+        auto newParams = parseParamOrLocal(curr);
+        params.insert(params.end(), newParams.begin(), newParams.end());
+      } else if (elementStartsWith(curr, RESULT)) {
+        auto newResults = parseResults(curr);
+        results.insert(results.end(), newResults.begin(), newResults.end());
+      }
+    }
+    return Signature(Type(params), Type(results));
+  } else if (*s[0] == STRUCT) {
+    FieldList fields;
+    for (size_t k = 1; k < s.size(); k++) {
+      auto& t = *s[k];
+      bool mutable_ = false;
+      if (t.isStr()) {
+        // t is a simple string name like "i32"
+        fields.emplace_back(elementToType(t), mutable_);
+        continue;
+      }
+      // t is a tuple, containing either
+      //   (field TYPE)
+      // or
+      //   (field $name TYPE)
+      if (*t[0] != FIELD) {
+        throw ParseException("invalid struct field", s.line, s.col);
+      }
+      if (t.size() != 2 && t.size() != 3) {
+        throw ParseException("invalid field size", s.line, s.col);
+      }
+      if (t.size() == 3) {
+        if (!t[1]->isStr()) {
+          throw ParseException("invalid field name", s.line, s.col);
+        }
+        // TODO: save the name of the field.
+      }
+      Element* last = t[t.size() - 1];
+      // The last element may also be (mut (..)).
+      if (last->isList() && *(*last)[0] == MUT) {
+        mutable_ = true;
+        last = (*last)[1];
+      }
+      fields.emplace_back(elementToType(*last), mutable_);
     }
+    return Struct(fields);
   }
-  return Signature(Type(params), Type(results));
+  throw ParseException("invalid heap type", s.line, s.col);
```

Comment:
I think by taking this struct and array parsing part out to the next PR (+ and `getFeatures` function change in wasm-type.cpp), this PR will become purely NFC refactoring and will not be blocked by anything else, and we can land this array/struct parsing hopefully with tests. I think that would make this PR simpler, WDYT?

Reply:
I agree it could be simpler that way. Happy to refactor it if you prefer. However, the benefit to doing it this way is that it is actually testable. This is, I think, the minimal amount of code to be testable. (But a test is not added, since adding a new test mode temporarily seems a bad idea.)

Reply:
I removed the struct parsing now.